src/platforms/common/swdptap.c

```diff
@@ -89,31 +99,49 @@ static uint32_t swdptap_seq_in_clk_delay(size_t clock_cycles) __attribute__((opt
 static uint32_t swdptap_seq_in_clk_delay(const size_t clock_cycles)
 {
 	uint32_t value = 0;
-	for (size_t cycle = 0; cycle < clock_cycles; ++cycle) {
-		gpio_clear(SWCLK_PORT, SWCLK_PIN);
-		value |= gpio_get(SWDIO_IN_PORT, SWDIO_IN_PIN) ? 1U << cycle : 0U;
+	if (!clock_cycles)
+		return 0;
+	for (size_t cycle = clock_cycles; cycle--;) {
```

Comment:
We're still not sold on turning this into a count-down loop due to the increased difficulty in reasoning about the termination of the loop - given the shift at the end, and how Thumb works, we'd imagine that testing `if (clock_cycles)` before the right shift and leaving this as a count up loop would result in the same or more efficient code, while more clearly conveying the termination conditions.

Reply:
I documented the difference in the commit message "adjust timing". Unfortunately, I didn't note the compiler version involved, which might be relevant.

Reply:
Could the reasoning be noted in the code please - if some other PR touches these lines and such, the blame log won't point to your commit, and trying to traverse it just to learn the intent and reasoning is a big non-starter. If it's inline as comments in the code, there's no complex discovery required.

We can then have an experiment in Compiler Explorer with a few different versions of GCC to see if it's compiler version specific and what the impacts are.

Reply:
Added comments. Please let me know if you'd prefer that I deduplicate some of them and replace them with references.